toolchain/sem_ir/builtin_function_kind.cpp

```diff
@@ -164,29 +164,53 @@ constexpr BuiltinInfo FloatMakeType = {"float.make_type",
 constexpr BuiltinInfo BoolMakeType = {"bool.make_type",
                                       ValidateSignature<auto()->Type>};
 
-// "int.negate": integer negation.
-constexpr BuiltinInfo IntNegate = {"int.negate",
-                                   ValidateSignature<auto(IntT)->IntT>};
+// "int.snegate": integer negation.
+constexpr BuiltinInfo IntSNegate = {"int.snegate",
+                                    ValidateSignature<auto(IntT)->IntT>};
```

Comment:
Yeah, I think we wanted to mirror the names that LLVM uses here, hence `sadd` / `uadd` matching `APInt`. But these aren't public API so I don't really care too much, and I'm fine with reconsidering that if these names are making you sadd.

I think we could just about get rid of `make_type_signed` / `make_type_unsigned` after all the builtins are changed to not care about the signedness of the type -- we could have a single `int.make_type(N)` builtin that produces a "sequence of N bits" type, like LLVM's `iN`. The remaining issue, I think, is that we use the signedness in diagnostics and SemIR to determine how to print values of the type to the user. That too could be fixed if we use an interface to determine how to print values for diagnostics; I don't know if that's something we would want to explore in the short term, but we could consider that as a long-term approach here. In the short term, I'm open to renaming the `make_type_*` functions to whatever's consistent with our other decisions.

I think I like keeping a consistent "int." prefix, given that I think it would be reasonable to expose the "unsigned" (wrapping) operations on `Core.Int(N)` with different names -- I think treating the choice as part of the operation rather than part of its type or namespace at the core level is likely to be a good thing.

One possible option here would be to focus on the meaning of the operations rather than on the "signedness" -- eg, `int.add` (signed overflow is an error) vs `int.add.wrap` (overflow wraps). That would work for `add`, `sub`, `mul`, and `negate`, but not for `div`, `mod`, and comparisons.

We also need to decide how to name the right-shift operations. I'd been planning on `int.ashr` (arithmetic right shift) vs `int.lshr` (logical right shift) to match LLVM, but if we deviate from LLVM naming then we could call those "signed" vs "unsigned" if that's more consistent.

Reply:
I was expecting the `make_type` calls to want to return something semi-typed. The return value needs to work with Carbon's type system (i.e., `i32` versus `u32` cannot be distinguished based on bit width).

But I also think it's temporary: once we have a native `Int(32)` typing capacity, wouldn't we just have that contain storage such as `bits.allocate(N)`?

FWIW I lean towards distinguishing builtins based on how Carbon plans to use them, rather than how LLVM IR defines them, if we're only exposing these for the core library. In part there, I expect that things that are more broadly exposed, as in for end users, would probably have a different API.

Reply:
We can't just use something like `bits.allocate(N)` -- we need enough typing in the type primitives to know how to represent the type in LLVM IR (eg, `iN` versus any other N-bit type) and how to represent it during constant evaluation / how to type-check the inputs to the builtin functions. We still need somewhat high-level primitive types for both purposes (eg, "this is a primitive N-bit integer", not just "this is N bits wide"), but we should be able to get to the point where we have something like:

```carbon
package Core api;

fn MakeIntN(N:! BigInt) -> type = "int.make";

class Int(N:! BigInt) {
  adapt MakeIntN(N);

  impl as Add {
    fn Op[self: Self](other: Self) = "int.sadd";
  }
}

class UInt(N:! BigInt) {
  adapt MakeIntN(N);

  impl as Add {
    fn Op[self: Self](other: Self) = "int.uadd";
  }
}
```

Reply:
Wouldn't the casting behavior of llvm types be sufficient? I had thoughts types were mainly aesthetic, that these were actually always force-cast in essence?

That is, the builtin function could partly be defining how to convert parameters for llvm's purposes. I'm not sure it needs to be defined by the result of "int.make".

Reply:
The pointee types of pointers don't matter to LLVM, but types do otherwise matter quite a lot. Maybe the biggest thing is that the type used to represent function parameters and return values affects the calling convention used for the function. For local values, if we get the type "wrong" we'll end up at best inserting a lot of `bitcast` instructions in the IR we generate, but that's not always possible -- I think the most obvious representation for "N bytes of memory" would be `[i8 x N]`, which is not even `bitcast`able and would require going through memory.

We need our types to specify their representation enough that we get the LLVM IR types we want; I don't think there's any way around that. I think the idea of using a `bits.allocate(N)` for everything, while appealing, won't work. We can probably do something like `allocate(Size, Alignment, LLVM type)` -- but even then, I think having stronger typing will make the constant evaluation side of things much simpler for us to deal with, and give us the ability to evaluate more efficiently.

Reply:
I'm going to go ahead and land this split of the builtins; we can iterate on the naming separately, based on the results of the discussion on discord.